source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -466,19 +466,19 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
     if (!cluster) {
       return false;
     }
-    // TODO(snowp): Ideally we should prevent this from being configured, but that's tricky to get
-    // right since whether a cluster is invalid depends on both the tcp_proxy config + cluster
-    // config.
-    if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) == 0) {
-      ENVOY_LOG(error, "Attempted to tunnel over HTTP/1.1, this is not supported. Set "
-                       "http2_protocol_options on the cluster.");
-      return false;
-    }
     Http::ConnectionPool::Instance* conn_pool = cluster_manager_.httpConnPoolForCluster(
         cluster_name, Upstream::ResourcePriority::Default, absl::nullopt, this);
     if (conn_pool) {
-      upstream_ = std::make_unique<HttpUpstream>(*upstream_callbacks_,
-                                                 config_->tunnelingConfig()->hostname());
+      if ((cluster->info()->features() & Upstream::ClusterInfo::Features::HTTP2) == 0) {
+        // TODO(irozzo): How to detect HTTP3
```

Comment:
I think you can replace the old early return-false with features() & HTTP3 != 0

Reply:
Unfortunately, as we discussed offline the only feature flag I see at the moment is about HTTP2: https://github.com/envoyproxy/envoy/blob/master/include/envoy/upstream/upstream.h#L718.
Anyway, HTTP3 upstreams do not seem to be supported yet.